src/main/scala/mesosphere/marathon/core/storage/impl/cache/LazyCachingPersistenceStore.scala

```diff
@@ -0,0 +1,142 @@
+
+/**
+  * A Write Ahead Cache of another persistence store that lazily loads values into the cache.
+  *
+  * @param store The store to cache
+  * @param mat a materializer for Akka Streaming
+  * @param ctx The execution context for future chaining.
+  * @tparam K The persistence store's primary key type
+  * @tparam Serialized The serialized format for the persistence store.
+  */
+class LazyCachingPersistenceStore[K, Category, Serialized](
+    store: BasePersistenceStore[K, Category, Serialized])(implicit
+  mat: ActorMaterializer,
+    ctx: ExecutionContext) extends PersistenceStore[K, Category, Serialized] with StrictLogging {
+
+  private[storage] val idCache = RwLock(TrieMap[Category, Seq[Any]]())
+  private[storage] val valueCache = RwLock(TrieMap[K, Option[Any]]())
+
+  override def ids[Id, V]()(implicit ir: IdResolver[Id, V, Category, K]): Source[Id, NotUsed] = {
+    val storeId = ir.category
+    idCache.read { ids =>
+      if (ids.contains(storeId)) {
+        Source(ids(storeId).map(_.asInstanceOf[Id]))
+      } else {
+        val childFuture = async {
+          val children = await(store.ids.toMat(Sink.seq)(Keep.right).run())
+          idCache.write(c => c(storeId) = children)
+          children
+        }
+        Source.fromFuture(childFuture).mapConcat(identity)
+      }
+    }
+  }
+
+  override def deleteAll[Id, V](k: Id)(implicit ir: IdResolver[Id, V, Category, K]): Future[Done] = {
+    async {
+      await(store.deleteAll(k))
+      val storageId = ir.toStorageId(k, None)
+      valueCache.write { vc =>
+        vc.remove(storageId)
+        idCache.write { c =>
+          val old = c.getOrElse(ir.category, Nil)
+          val children = old.filter(_ != storageId)
+          if (children.nonEmpty) {
+            c.put(ir.category, children)
+          } else {
+            c.remove(ir.category)
+          }
+        }
+      }
+      Done
+    }
+  }
+
+  override def get[Id, V](id: Id)(implicit
+    ir: IdResolver[Id, V, Category, K],
+    um: Unmarshaller[Serialized, V]): Future[Option[V]] = {
+    val storageId = ir.toStorageId(id, None)
+    val cached = valueCache.read(_.get(storageId))
+    cached match {
+      case Some(v) =>
+        Future.successful(v.asInstanceOf[Option[V]])
+      case None =>
+        async {
+          val value = await(store.get(id))
+          valueCache.write(_.put(storageId, value))
+          value
+        }
+    }
+  }
+
+  override def get[Id, V](id: Id, version: OffsetDateTime)(implicit
+    ir: IdResolver[Id, V, Category, K],
+    um: Unmarshaller[Serialized, V]): Future[Option[V]] =
+    store.get(id, version)
+
+  override def store[Id, V](id: Id, v: V)(implicit
+    ir: IdResolver[Id, V, Category, K],
+    m: Marshaller[V, Serialized],
+    um: Unmarshaller[Serialized, V]): Future[Done] = {
+    async {
+      await(store.store(id, v))
+      val storageId = ir.toStorageId(id, None)
+      val storageCategory = ir.category
+      valueCache.write { values =>
```

Comment:
"the main thing to understand about async/await is that it's just another way of doing Future's flatMap. there's no concurrency magic behind the scenes. all the calls inside an async block are sequential, including await which doesn't actually block the executing thread but rather wraps the rest of the async block in a closure and passes it as a callback on completion of the Future we're waiting on. so in the first piece of code the second calculation doesn't start until the first await has completed since no one started it prior to that."

So, this should work as expected no? I think I can add a little more parallelism elsewhere by having things like:

val storeCurrent = rawStore()
val storeAsVersioned = rawStore()
(await(storeCurrent), await(storeAsVersioned))